### ISSUE TITLE
check-expect stops evaluating tests after errors

### ISSUE THREAD

**Issue by Blaisorblade** [racket/htdp#19]
I have a code snippet with two failing `check-expect` tests, where the second failure is hidden by the first failing — test evaluation simply stops silently, without warnings. I argue there's a bug (at least usability-wise):
- I don't see that even documented (in, e.g., http://docs.racket-lang.org/test-engine/).
- I'm not convinced that's appropriate, except for critical errors, but one can argue about it. In particular, there are downsides to resuming testing after failure where mutations and side effects are involved, for tests with accidental effect leaks, but such tests are bad even when they're possible (in the advanced student languages), and they're impossible in BSL and ISL.
- More importantly, it's clearly important to warn users clearly that later tests weren't run, in case one assumes otherwise. That's especially easy since tests are first registered and then run, so it should be easy to check the total test count.

This is distilled from a first-year student problem: she reached the conclusion (through otherwise sensible deductions) that the second test was working, when it wasn't even run. Overall, she appears here an intelligent student who got stuck, also because of confusing tools. Since the tools confused me for a while too, I think this is a clear bug.
The examples below are not as confusing, but just because I've oversimplified the tests during minimization.

``` racket
(define (f x)
  (cond
    [#false 0]))
(check-expect (f 0) 0)
(check-expect 1 0)
```

```
1 Test gelaufen.
0 Tests bestanden.
Keine Signaturverletzungen.

Check-Fehler:
    check-expect bekam den folgenden Fehler statt des erwarteten Werts, 5.
   :: cond: all question results were false
in Zeile 4, Spalte 0
```

Here you can notice that the Racket only reports that one test run, so here the problem is obvious. But my student had 40 tests, and I noticed by sheer luck: I added a failing test, the test count didn’t change, and I happened to be paying attention because she had reported something was amiss.

**Comment by uchwings** [comment]
I agree this is highly undesirable behavior; it seems like errors should be considered mere local test failures, and if they're so bad as to merit aborting the entire test runner, this should be noted and warned about in the output. Silently not running all specified tests is not acceptable.

**Comment by uchwings** [comment]
It looks like this behavior can be changed by editing `test-engine/racket-tests.rkt` and under the definition of `run-and-check`, changing `(if exn (raise exn) #f)` to simply `#f`.

**Comment by jbclements** [comment]
I think you should make a pull request for this, if you have time. I'm not totally convinced that it's the right thing to do, but maybe it is.

**Comment by Blaisorblade** [comment]
Thanks for looking into this. Can you turn it into a PR?
I’m not running an HtDP course currently but colleagues are (@b-studios, @phischu and @lordxist), and might be interested in testing the fix.

**Comment by Blaisorblade** [comment]
@jbclements As @uchwings says, the alternative would be a *verbose* abort. In fact, that should also be documented, possibly even in HtDP if that explains `check-expect` (ideally, HtDP should not contain accidental lies; conscious oversimplifications are a different matter).

I’ve TAed HtDP in 2014/‘15 so I forgot a lot, but I’d argue errors should be distinguished in the output *if and only if* HtDP teaches about errors (which I think it does).

**Comment by samth** [comment]
This was fixed, see #50 for discussion.
